### PR TITLE
Changes to cluster autoscaling

### PR DESCRIPTION
**Pull request by mattmb**
This is a WIP (needs testing, tests and docs etc.)

Just putting it up for early comments on the direction.

Also needs https://github.com/Yelp/paasta/pull/521/ to make the draining work and (I think) further changes to the maintenance scripts to prevent allocation of tasks to slaves in maintenance mode.

### CODE REVIEW

**Comment by mattmb** [comment]
Should be possible to support ASGs too. We'd just have to detach the instance before terminating: http://docs.aws.amazon.com/cli/latest/reference/autoscaling/detach-instances.html

File: paasta_tools/autoscaling_lib.py [hunk 1]
```diff
@@ -508,41 +512,130 @@ def spotfleet_scaler(resource, error):
...
+def slave_pid_to_ip(slave_pid):
+    """Convert slave_pid to IP
+
+    :param: slave pid e.g. slave(1)@10.40.31.172:5051
+    :returns: ip address"""
+    regex = re.compile(r'.+?@([\d\.]+):\d+')
+    return regex.match(slave_pid).group(1)
+
+
+def pick_slaves_to_kill(mesos_state, number, pool='default'):
```
Comment (solarkennedy) [inline paasta_tools/autoscaling_lib.py]: I think longer term this function has to have the actual list from the the SFR/ASG in order to work.

Otherwise it won't take into account instances that are not even registered in mesos (duds, broken puppet, etc)

File: paasta_tools/autoscaling_lib.py [hunk 2]
```diff
@@ -508,41 +512,130 @@ def spotfleet_scaler(resource, error):
...
     pass
 
 
+def get_instance_id_from_ip(ip):
+    """Get AWS instance ID from private IP
+
+    :param ip: private IP of AWS instance
+    :returns: Instance ID"""
+    ec2_client = boto3.client('ec2')
+    instances = ec2_client.describe_instances(Filters=[{'Name': 'private-ip-address', 'Values': [ip]}])
```
Comment (solarkennedy) [inline paasta_tools/autoscaling_lib.py]: Hmmmm.

I wonder if this only counts as "1" api call....

I'm pretty sure we are going to do something about that soon :(

File: paasta_tools/mesos_tools.py [hunk 3]
```diff
@@ -615,3 +615,36 @@ def get_mesos_network_for_net(net):
         'host': 'HOST',
     }
     return docker_mesos_net_mapping.get(net, net)
+
+
+def get_mesos_task_count_by_slave(mesos_state, pool='default'):
+    """Get counts of running tasks per mesos slave. Also include separate count of chronos tasks
```
Comment (solarkennedy) [inline paasta_tools/mesos_tools.py]: why do we actually need a separate chronos_count?

are you intending to eventually weight them differently?

File: paasta_tools/autoscaling_lib.py (hunk 1 above)
Comment (mattmb) [inline paasta_tools/autoscaling_lib.py]: Hmm, this is a good point we definitely want to clear up anything that's not registered in mesos!

I was trying to keep this function agnostic of the autoscaling provider but maybe I can add something somewhere else to achieve this

File: paasta_tools/autoscaling_lib.py (hunk 2 above)
Comment (mattmb) [inline paasta_tools/autoscaling_lib.py]: You're worried we might hit some rate limits? I was also thinking that this was a bit convoluted, is there some way we can store the instance ID in mesos when we bootstrap the slaves?

File: paasta_tools/mesos_tools.py (hunk 3 above)
Comment (mattmb) [inline paasta_tools/mesos_tools.py]: So my thinking was that batches are more difficult to "drain". In the sense that the best you can do is wait for the current run to finish. So I'm trying to sort by `chronos_count` and then the total number of tasks.

File: paasta_tools/autoscaling_lib.py [hunk 4]
```diff
@@ -508,41 +512,130 @@ def spotfleet_scaler(resource, error):
...
+    :param instances_to_kill: dict of instances to kill {<slave_id>: <instance_id>}"""
+    ec2_client = boto3.client('ec2')
+    try:
+        with Timeout(seconds=CLUSTER_DRAIN_TIMEOUT):
+            while instances_to_kill:
+                mesos_state = get_mesos_state_from_leader()
+                slaves_count = get_mesos_task_count_by_slave(mesos_state, pool=None)
+                for slave_id, instance_id in instances_to_kill.items():
+                    num_tasks = slaves_count[slave_id]['count']
+                    log.debug("Instance: {0}, still running {1} tasks".format(instance_id, num_tasks))
```
Comment (Rob-Johnson) [inline paasta_tools/autoscaling_lib.py]: drop the 'still' here, and can you add a debug line for when you terminate please

File: paasta_tools/autoscaling_lib.py [hunk 5]
```diff
@@ -37,18 +38,24 @@
 from paasta_tools.marathon_tools import MESOS_TASK_SPACER
 from paasta_tools.marathon_tools import set_instances_for_marathon_service
 from paasta_tools.mesos_tools import get_mesos_state_from_leader
+from paasta_tools.mesos_tools import get_mesos_task_count_by_slave
 from paasta_tools.mesos_tools import get_running_tasks_from_active_frameworks
+from paasta_tools.paasta_maintenance import drain
```
Comment (Rob-Johnson) [inline paasta_tools/autoscaling_lib.py]: this needs moving out into a more generic 'maintenance' library. ``paasta_maintenance`` is a script, and we shouldn't be importing from it here.

File: paasta_tools/autoscaling_lib.py [hunk 6]
```diff
@@ -508,41 +512,130 @@ def spotfleet_scaler(resource, error):
...
+    sfr_id = resource['id']
+    ec2_client = boto3.client('ec2')
+    if delta == 0:
+        return
+    elif delta > 0:
+        return ec2_client.modify_spot_fleet_request(SpotFleetRequestId=sfr_id, TargetCapacity=target_capacity,
+                                                    ExcessCapacityTerminationPolicy='noTermination')
+    elif delta < 0:
+        slaves_by_ip = {slave: slave_pid_to_ip(slave) for slave in slaves_to_kill}
+        instances_to_kill = {slave_id: get_instance_id_from_ip(ip) for slave_id, ip in slaves_by_ip.items()}
```
Comment (Rob-Johnson) [inline paasta_tools/autoscaling_lib.py]: I don't see you catching the IndexError that this function throws

File: paasta_tools/autoscaling_lib.py [hunk 7]
```diff
@@ -508,41 +512,130 @@ def spotfleet_scaler(resource, error):
...
+        start = time.time()
+        # This ensures we are in maintenance mode longer than the time before
+        # we give up trying to drain and just terminate
+        duration = CLUSTER_DRAIN_TIMEOUT + 300
+        drain(slaves_by_ip.values(), start, duration)
```
Comment (Rob-Johnson) [inline paasta_tools/autoscaling_lib.py]: please modify this function to return the http response and let the caller decide what to do, rather than relying on it's side-effecty behaviour

File: paasta_tools/mesos_tools.py [hunk 8]
```diff
@@ -615,3 +615,36 @@ def get_mesos_network_for_net(net):
...
+
+    :param mesos_state: mesos state dict
+    :param pool: pool of slaves to return (None means all)
+    :returns: dict of {<slave_id>:{'count': <count>, 'slave': <slave_object>, 'chronos_count':<chronos_count>}}
+    """
+    all_mesos_tasks = get_running_tasks_from_active_frameworks('')  # empty string matches all app ids
+    if pool:
```
Comment (Rob-Johnson) [inline paasta_tools/mesos_tools.py]: so we're defaulting pool to 'default' here - is the only way to trigger this by setting pool=None, in which case we'll be given the default pool anyway? Can you clarify what case this is covering?

File: paasta_tools/mesos_tools.py [hunk 9]
```diff
@@ -615,3 +615,36 @@ def get_mesos_network_for_net(net):
...
+        slaves = {
+            slave['id']: {'count': 0, 'slave': slave, 'chronos_count': 0} for slave in mesos_state.get('slaves', [])
+            if slave['attributes'].get('pool', 'default') == pool
+        }
+    else:
+        slaves = {
+            slave['id']: {'count': 0, 'slave': slave, 'chronos_count': 0} for slave in mesos_state.get('slaves', [])
```
Comment (Rob-Johnson) [inline paasta_tools/mesos_tools.py]: can you create a NamedTuple to represent this dict please, to make the structure of it clearer?

File: paasta_tools/autoscaling_lib.py [hunk 10]
```diff
@@ -508,41 +512,130 @@ def spotfleet_scaler(resource, error):
...
+            while instances_to_kill:
+                mesos_state = get_mesos_state_from_leader()
+                slaves_count = get_mesos_task_count_by_slave(mesos_state, pool=None)
+                for slave_id, instance_id in instances_to_kill.items():
+                    num_tasks = slaves_count[slave_id]['count']
+                    log.debug("Instance: {0}, still running {1} tasks".format(instance_id, num_tasks))
+                    if num_tasks == 0:
+                        ec2_client.termintate_instances(InstanceIds=[instance_id])
+                        instances_to_kill.pop(slave_id)
+                time.sleep(5)
```
Comment (Rob-Johnson) [inline paasta_tools/autoscaling_lib.py]: please add a log note here to warn people that you're sleeping

File: paasta_tools/autoscaling_lib.py [hunk 11]
```diff
@@ -508,41 +512,130 @@ def spotfleet_scaler(resource, error):
 
     if new_capacity != current_capacity:
         print "Scaling SFR %s from %d to %d!" % (resource['id'], current_capacity, new_capacity)
-        scale_aws_spot_fleet_request(sfr_id=resource['id'], target_capacity=new_capacity, ec2_client=ec2_client)
+        return new_capacity - current_capacity, new_capacity
+    else:
+        return 0
+
+
+def slave_pid_to_ip(slave_pid):
```
Comment (Rob-Johnson) [inline paasta_tools/autoscaling_lib.py]: can you move this to ``mesos_tools`` please

File: paasta_tools/mesos_tools.py [hunk 12]
```diff
@@ -615,3 +615,36 @@ def get_mesos_network_for_net(net):
...
+        }
+    for task in all_mesos_tasks:
+        if task.slave['id'] not in slaves:
+            log.error("Task associated with %s but slave not in masters list of slaves" % task.slave['id'])
+            continue
+        else:
+            slaves[task.slave['id']]['count'] += 1
+            log.debug("Task framework: {0}".format(task.framework.name))
+            if task.framework.name == 'chronos':
```
Comment (Rob-Johnson) [inline paasta_tools/mesos_tools.py]: Can you verify this is correct? I thought chronos registers as ``chronos-2.4.0`` in mesos.

**Comment by Rob-Johnson** [comment]
Please add some tests.

File: paasta_tools/mesos_tools.py (hunk 12 above)
Comment (mattmb) [inline paasta_tools/mesos_tools.py]: Ugh, you're right! As discussed, I'm going to tackle this with a puppet change and a PR against Chronos to remove the version from this property rather than make this any more hacky here.

File: paasta_tools/mesos_tools.py (hunk 8 above)
Comment (mattmb) [inline paasta_tools/mesos_tools.py]: I put this in so that there's a way to get *all* the pools. I'm not using it for the autoscaler but I think I'll leave it as it could be useful elsewhere.

File: paasta_tools/mesos_tools.py (hunk 9 above)
Comment (mattmb) [inline paasta_tools/mesos_tools.py]: Don't think I can use a tuple because I need to update the counts below?

**Comment by mattmb** [comment]
Ok, pretty happy with this now. Can someone take another look? Probably should write some documentation on the cluster level autoscaling but I'd suggest we get this deployed and start playing with it first.

File: paasta_tools/autoscaling_lib.py [hunk 13]
```diff
@@ -514,41 +547,165 @@ def spotfleet_scaler(resource, error):
...
+    ec2_client = boto3.client('ec2')
+    try:
+        # This loop should always finish because the maintenance window should trigger is_ready_to_kill
+        # being true. Just in case though we set a timeout and terminate anyway
+        with Timeout(seconds=CLUSTER_DRAIN_TIMEOUT + 300):
+            while instances_to_kill:
+                for slave_pid, slave_data in instances_to_kill.items():
+                    instance_id = slave_data['instance_id']
+                    if not instance_id:
+                        # if we failed to find instance id then skip
```
Comment (solarkennedy) [inline paasta_tools/autoscaling_lib.py]: Just log.debug instead of a comment here

File: paasta_tools/autoscaling_lib.py [hunk 14]
```diff
@@ -514,41 +547,165 @@ def spotfleet_scaler(resource, error):
...
+                        # shouldn't happen but just incase
+                        continue
+                    is_ready_to_kill = True
```
Comment (solarkennedy) [inline paasta_tools/autoscaling_lib.py]: I would like to see an optimization here eventually where

    is_ready_to_kill or has_no_tasks or is_not_in_mesos

File: paasta_tools/autoscaling_lib.py [hunk 15]
```diff
@@ -514,41 +547,165 @@ def spotfleet_scaler(resource, error):
...
+        start = int(time.time() + CLUSTER_DRAIN_TIMEOUT)
+        # Set the duration to an hour, if we haven't cleaned up and termintated by then
+        # mesos should put the slave back into the pool
+        duration = 600
+        log.debug("Draining {0}".format(instances_to_kill))
+        log.debug("Decreasing spot fleet capacity to: {0}".format(target_capacity))
+        if not dry_run:
+            # sort to make testing easier
+            drain([instance['ip'] for instance in sorted(instances_to_kill.values())], start, duration)
+            ec2_client.modify_spot_fleet_request(SpotFleetRequestId=sfr_id, TargetCapacity=target_capacity,
```
Comment (solarkennedy) [inline paasta_tools/autoscaling_lib.py]: I'm a little concerned about the method here.

What happens if something goes wrong with the `wait_and_terminate` step? I don't think the autoscaler will be able to recover because those instances will be "lost".

I think you have to do it "atomically" where scale + kill at the same time on each instance, otherwise you risk instances falling through the cracks

File: paasta_tools/mesos_tools.py (hunk 8 above)
Comment (Rob-Johnson) [inline paasta_tools/mesos_tools.py]: right, but what I mean is the default value doesn't make sense as it currently stands. If I wanted to get **all** the pools, then I'd have to do ``get_mesos_task_count_by_slave(mesos_state, pool=None)``, which is odd.


Two suggestions here:

- Set the default parameter to be None, so that you get **all** pools unless you specify otherwise
```
def get_mesos_task_count_by_slave(mesos_state, pool=None)
```


- rename the parameter to ``pool_filter``, and have it as a function that returned ``True`` when you matched a pool that you're interested in, and ``False`` otherwise. The caller could then do something like:
```
get_mesos_task_count_by_slave(mesos_state, lambda pool: pool == 'default')
get_mesos_task_count_by_slave(mesos_state, lambda pool: pool == 'lucy')
get_mesos_task_count_by_slave(mesos_state, lambda pool: True) #match any
```

I kind of like the second one for it's flexibility, but I'll leave it with you to decide.

File: paasta_tools/autoscaling_lib.py (hunk 14 above)
Comment (mattmb) [inline paasta_tools/autoscaling_lib.py]: Agreed. So my initial plan is for the call to paasta maintenance to check the slave is running no tasks OR it has reached the maintenance window (which means it failed to drain in a timely fasion).

I'm thinking I might handle the `is_not_in_mesos` earlier on because by this point we're only looking at slaves that we found.

File: paasta_tools/mesos_tools.py (hunk 9 above)
Comment (Rob-Johnson) [inline paasta_tools/mesos_tools.py]: just make the return value a namedtuple then - it's a real pain to have to figure out what keys might be in the dict returned everyime you want to use a function. a NamedTuple puts a contract around that.

File: paasta_tools/autoscaling_lib.py [hunk 16]
```diff
@@ -514,41 +547,165 @@ def spotfleet_scaler(resource, error):
...
+        if not dry_run:
+            ec2_client.modify_spot_fleet_request(SpotFleetRequestId=sfr_id, TargetCapacity=target_capacity,
+                                                 ExcessCapacityTerminationPolicy='noTermination')
+            return
+    elif delta < 0:
+        number_to_kill = delta * -1
+        sfr_ips = get_sfr_instance_ips(sfr_id)
+        log.debug("IPs in SFR: {0}".format(sfr_ips))
+        sfr_sorted_slaves = [slave for slave in sorted_slaves if slave_pid_to_ip(slave['pid']) in sfr_ips]
+        log.debug("SFR slave kill preference: {0}".format(sfr_sorted_slaves))
```
Comment (Rob-Johnson) [inline paasta_tools/autoscaling_lib.py]: can you just print the slave ids here, rather than the entire object? this log line is going to be very hard to parse otherwise

File: paasta_tools/autoscaling_lib.py [hunk 17]
```diff
@@ -514,41 +547,165 @@ def spotfleet_scaler(resource, error):
...
+        if number_to_kill > len(sfr_sorted_slaves):
+            log.error("Didn't find enough candidates to kill. This shouldn't happen so let's not kill anything!")
+            return
+        else:
```
Comment (Rob-Johnson) [inline paasta_tools/autoscaling_lib.py]: you don't need the else clause here.

File: paasta_tools/autoscaling_lib.py [hunk 18]
```diff
@@ -514,41 +547,165 @@ def spotfleet_scaler(resource, error):
...
+            slaves_to_kill = sfr_sorted_slaves[0:number_to_kill]
+        log.debug("Set to kill: {0}".format([slave['pid'] for slave in slaves_to_kill]))
+        instances_to_kill = {}
+        for slave in slaves_to_kill:
+            ip = slave_pid_to_ip(slave['pid'])
+            instances_to_kill[slave['pid']] = {'ip': ip}
```
Comment (Rob-Johnson) [inline paasta_tools/autoscaling_lib.py]: can you reduce this into one line:

```
instances_to_kill[slave['pid']] = {
    'ip': ip,
    'instance_id': get_instance_id_from_ip(ip)
}
```

File: paasta_tools/autoscaling_lib.py (hunk 15 above)
Comment (mattmb) [inline paasta_tools/autoscaling_lib.py]: That's a good point. I guess I could reduce target capacity by 1, then wait_and_terminate and repeat until we reach the target.

I suppose the only downside is if a particular instance takes a while to drain then we make the whole scaling event take longer? But I guess we don't want two runs to overlap anyway and we can be aggressive with the timeout on draining if need be.

File: paasta_tools/mesos_tools.py (hunk 9 above)
Comment (mattmb) [inline paasta_tools/mesos_tools.py]: Fair point, I guess it makes it easier to consume. I'll fixup :-)

File: paasta_tools/mesos_tools.py (hunk 8 above)
Comment (mattmb) [inline paasta_tools/mesos_tools.py]: Right, I see what you mean now 👍 

In the interests of keeping it simple I'm going to do option 1.

**Comment by mattmb** [comment]
After discussion, I agree with @solarkennedy that we need to be more atomic when we scale down (especially so we handle the case when we fail to clean up an instance). We also need to add some logic to the scale down to take account of instance weightings.

My plan is to ship this and get it working in "dry_run" mode. Then I'll tackle the remaining points in a fresh branch.

@Rob-Johnson @solarkennedy are you happy with the other fixups? If so I'll autosquash and merge.

**Comment by solarkennedy** [comment]
Yep, I think this is one of those things where we pretty much won't get it right the first time, or heck even close, and we'll have to run it in noop mode and iterate.

Ship!

**Comment by Rob-Johnson** [comment]
ship!
